### ISSUE TITLE
bind mounts in fstab work improperly when pointing to ZFS

### ISSUE THREAD

**Issue by bhodgens** [zfsonlinux/zfs#971]
If using bind mounts from within fstab to a location within a zpool, the mount point will behave oddly, not listing the full/actual contents of the destination (just directories, from what I can tell - might this be due to fstab being parsed and the source getting mounted prior to zfs taking control, with an odd behavior resulting?)

If mounted via mount after the system has fully booted (eg. mount -o bind /home /zpool/home),  it behaves properly. 

To replicate, make an entry in fstab and reboot (or export the pool, run the mount, and reimport): 
/zpool/home   /home    none   bind,defaults   0 0

Distro: Ubuntu 12.04.1 
zfsonlinux: 0.6.0rc10 from PPA

I'm using the zfs-mount and zfs-share scripts, not the 'mountall' modified binary from the PPA (which appears to require the mount=legacy option?)

**Comment by jbnance** [comment]
Is there realistic potential for this issue to be addressed?  I'm in the same scenario - NFSv4 + bind mounts - and would very much like to have a reasonable, manageable solution.

**Comment by afontenot** [comment]
@jbnance IMO, the correct Linux solution to this problem is described at the bottom of this [article on the Arch wiki](https://wiki.archlinux.org/index.php/ZFS#Bind_mount). 

Basically, you want to add a mount option in your fstab that will cause the system to wait for the file system to be mounted.

    /mnt/zfspool	/music	/srv/nfs/music		none	bind,x-systemd.requires=zfs-mount.service	0 0

**Comment by jbnance** [comment]
@afontenot thanks for the info, I'll try it out.  If it works as described on other distros (especially EL7+) I would agree.  I've been using the `_netdev` option thus far to try and deal with this so the `x-systemd` stuff is a natural progression.

**Comment by neocogent** [comment]
I had this problem on a media nas box and the solution suggest by @afontenot  worked perfectly. My nfs shares could not reach into the zpool without manually "fixing" after boot but adding the mount option gave me a clean boot that worked fine first time. 

Adding a mount option (x-systemd.requires=zfs-mount.service) seems to me a simple and sensible solution for this issue. It probably just needs to be documented and better known.

**Comment by tfgm-bud** [comment]
@afontenot, Just chiming in to say, the approach of changing the bind mounts in `/etc/fstab` worked great for me on Ubuntu 16.04 with the HWE kernel (https://wiki.ubuntu.com/Kernel/LTSEnablementStack).  Not sure if the HWE kernel is required but highly recommended as the ZFS version as of today on the HWE is v0.7.5.

**Comment by gmelikov** [comment]
I prefer to close this issue, 0.8 will have a better systemd integration, and there is a workaround https://github.com/zfsonlinux/zfs/issues/971#issuecomment-325165200

**Comment by XLTechie** [comment]
Still having this problem in 2026. Found this after running into this issue in Ubuntu 22.04, kernel 5.15.0-168.

I host a bunch of websites with large datasets, which are each mounted in /home/www/site, which is bind-mounted on a ZFS pool.
Realized recently that they were all showing up as mounted twice: once over /dev/sda (the root FS), and once over their proper location.
After rebooting from one year of operations, they all bind-mounted over the root FS, and I had to umount them, and `mount -a`, at which point they were all mounted correctly (I tested one in between, of course).
Obviously a late mount of underlying ZFS problem.
**Thanks to @afontenot for the working solution!**
